src/ui/wxWidgets/PasswordSafeFrame.cpp

```diff
@@ -1933,6 +2000,167 @@ CItemData* PasswordSafeFrame::GetBaseEntry(const CItemData *item) const
   return nullptr;
 }
 
+///////////////////////////////////////////////////////////////////////////////
+// TOTP Begin
+
+const CItemData* PasswordSafeFrame::GetTotpItem(const CItemData *item) const
+{
+  if (item == nullptr) {
+    // GetBaseEntry doesn't like nullptr
+    return nullptr;
+  }
+  if (item->IsNormal() || item->IsBase()) {
+    return item;
+  }
+  return m_core.GetBaseEntry(item);
+}
+
+bool PasswordSafeFrame::IsItemNormalOrBase(const CItemData *item) const
+{
+  return item == nullptr ? false : item->IsNormal() || item->IsBase();
+}
+
+bool PasswordSafeFrame::HasItemTwoFactorKey(const CItemData *item) const
+{
+  auto totpItem = GetTotpItem(item);
+  return (totpItem == nullptr) ? false : totpItem->IsTotpActive();
+}
+
+std::pair<StringX, StringX> PasswordSafeFrame::GetTotpData(const CItemData *item)
+{
+  if (item == nullptr) {
+    return std::make_pair(tostringx(wxT("n/a")), tostringx(wxT("n/a")));
+  }
+  StringX totp;
+  double ratio;
+  CItemData ciTemp(*GetTotpItem(item));
+  auto r = GetTwoFactorAuthenticationCode(ciTemp, totp, &ratio);
+  if (r != PWSTotp::Success) {
+    return std::make_pair(tostringx(wxT("------")), tostringx(wxT("--")));
```

Comment:
Why not use the same placeholder values here as on line 2032?

Reply:
The idea was to be able to distinguish between the different error cases.
Unified by using “n/a”. I would like to use a std::optional as return value to signal to the caller whether valid data is available or not, instead of doing this via the strings. Unfortunately, this is only supported as of C++17. However, a comment in the CMakeFile indicates a possible incompatibility with wxWidgets.